Document partial code examples

jdbaldry opened the pull request:
I'm starting to revise the documentation for formatting code examples and add an additional section for specifically handling partial code examples.

It turns out it's super hard!

I'd love any ideas on how we can better explain how to use snippets.
Is the increased verbosity avoidable or is it necessary to deal with complicated integration?


Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,110 @@ Additional tips:
+#### Configure syntax highlighting
 
-Introduce each code sample with a sentence or paragraph to establish its context. End the introduction with a colon if it immediately precedes the sample, or a period if it doesn't.
+In Markdown, an info string after the first three backticks is used to describe the language contained within.
```
```suggestion
In Markdown, use an info string after the first three backticks to describe the programming or markup language of the example.
```

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,110 @@ Additional tips:
 
-Introduce each code sample with a sentence or paragraph to establish its context. End the introduction with a colon if it immediately precedes the sample, or a period if it doesn't.
+In Markdown, an info string after the first three backticks is used to describe the language contained within.
+The website uses this information for syntax highlighting.
```
```suggestion
The website uses this information to apply syntax highlighting to code examples.
```

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,110 @@ Additional tips:
+In Markdown, an info string after the first three backticks is used to describe the language contained within.
+The website uses this information for syntax highlighting.
+
+To configure syntax highlighting for a JSON code example, refer to the following Markdown example:
```
This isn't really an example of how to **configure** anything, because in practice you don't need to wrap your code examples in markdown for them to work.  I wrote a couple of different revisions for this line, still not happy with this one...

```suggestion
The following is an example of how syntax highlighting is applied to a JSON code example:
```

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
`````diff
@@ -78,23 +78,110 @@ Additional tips:
+````
+
+Some common languages and their info strings are:
+
`````
```suggestion

- Bash: 'bash`
```

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,110 @@ Additional tips:
+#### Use the `code` shortcode for tabbed examples
+
+If you have the same example in multiple languages, use the [`code` shortcode](https://grafana.com/docs/writers-toolkit/write/shortcodes/#code).
+The website presents the snippets in tabs and remembers the users preferred choice.
```
```suggestion
The website presents the snippets in tabs and remembers the user's preferred choice.
```

JStickler commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,110 @@ Additional tips:
+> It is not a complete Service specification and must be incorporated with the rest of a Service specification.
+>
+> To incorporate the example, you must include it as the value to the Service `spec` mapping.
+> If there is an existing `ports` value, you must choose to replace it or merge the two.
```
Question, do we want to mention providing the dot notation (not sure if that's what it's officially called) with the path to where the snippet should live in the file, for example "The following sample would be inserted into your Kubernetes Service file under `spec.ports`."

jdbaldry replied:
I think we should for sure. I meant to add a comment to myself to that effect and clearly forgot :) 

I think it might require its own paragraph to explain the rules for the dot syntax.

JStickler replied:
Yeah, it might need a paragraph of explanation at that.

>It turns out it's super hard!
Truth!

jdbaldry commented:
Thanks for looking at this @jstickler. I took a while to get back to this PR but I've incorporated your suggestions in [`c69d7e2` (#348)](https://github.com/grafana/writers-toolkit/pull/348/commits/c69d7e2a2559847146e902ca8f5a5e3b77d58f53) and added a section on _dot notation_.

jdbaldry commented:
I'm also gonna have @imatwawana take a look since she wants to present this next week at the docs all hands.

imatwawana commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,135 @@ Additional tips:
+Use 2 spaces unless there is a clearly established alternative convention.
+For example, Python generally uses 4 spaces for indentation.
+
+Above all else, be consistent with other Grafana documentation.
```
What exactly do we mean here? Like if you're adding something, just make sure it's consistent with the other code samples around it? My confusion comes from the phrase "other Grafana documentation" because I interpret that as ALL other Grafana documentation.

josmperez replied:
I see Isabel's point. Is what we are trying to say "Above all else, be consistent with other Grafana documentation for code in the same language."?

jdbaldry replied:
Perhaps it is more confusing that it is valuable. I'm not sure exactly what I meant now that I have come back to this PR after some time.

I think it's kind of a hand-wavy "If your project is doing four tabs of indentation (WILD!) for all other code samples in this language, follow that style" but also "If Grafana engineering has a practice of doing X, then also do X."

Since this is the style guide, perhaps we should be more insistent that you do Y and if you are doing X, you must switch to doing Y.

imatwawana replied:
I think I'd say be consistent with existing documentation _in that area_ of the docs or in your project. I think we just want to be specific about what we're saying to be consistent with.

jdbaldry replied:
```suggestion
Above all else, be consistent with existing documentation in your area of the documentation or in your project.
```

imatwawana commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,135 @@ Additional tips:
+If you have the same example in multiple languages, use the [`code` shortcode](https://grafana.com/docs/writers-toolkit/write/shortcodes/#code).
+The website presents the snippets in tabs and remembers the user's preferred choice.
+
+### Format partial code examples
```
```suggestion
### Partial code examples
```

imatwawana commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,135 @@ Additional tips:
 For an in-depth, external resource about writing developer documentation, see [_Docs for Developers_](https://docsfordevelopers.com/).
 
-### Formatting code examples
+### Format code examples
```
I think starting these two formatting headings with the verb (especially since they aren't tasks) hides what we're trying to really get across here, which is one of these sections is about full code samples and one is about partial code samples. I've made the associated suggestion on line 137.
```suggestion
### Full code examples
```

imatwawana commented on docs/sources/write/style-guide/write-for-developers/index.md:
```diff
@@ -78,23 +78,135 @@ Additional tips:
+For example, `spec.template.metadata` to refer to the `metadata` field within the `template` field, which is itself within the `spec` field.
+
+If the field name has dots, surround the name with square brackets instead.
+For example, `spec.selector[app.kubernetes.io/name]`
```
```suggestion
For example, `spec.selector[app.kubernetes.io/name]`.
```
